---
title: "Vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Follow this commands to install the package:

```{r eval=FALSE}
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}
library(devtools)
devtools::install_github("https://github.com/LeonardoNossa/MotifRegressR")
```

```{r setup}
library(MotifRegressR)
```

## Description
This R package facilitates motif analysis for gene expression studies. It 
integrates upstream DNA sequence information, gene expression profiles, and 
Position Frequency Matrices (PFMs) to identify regulatory motifs influencing 
gene expression.

Users can import PFMs from JASPAR or custom sources, annotate genomic sequences 
with transcription start sites (TSS), and retrieve metadata of gene expression 
conditions.

The package scores motifs across genes, aggregates the results, and performs 
regression analysis to determine the most impactful motifs under different 
experimental conditions. Visualization functions finally help the user in the 
sense-making step of the analyses.


Below there is a visual scheme of the project.
If you are part of the University of Milan, you can also view it by clicking the following link and entering your credentials.

[whiteboard](https://unimi2013-my.sharepoint.com/:wb:/g/personal/marco_cominelli_studenti_unimi_it/EZApUGSaatFPlyUGVJopINMBtisiIJ_IU2fPwyi4fezndQ?e=mzdgal)

```{r, echo = FALSE, out.width = "680px"}
knitr::include_graphics(system.file("extdata", "whiteboard_screen", 
                                    package = "MotifRegressR"))
```

## Pipeline for the users in R
Here you can find the commands of the pipeline with added examples and 
explanations of the functions. In the examples, we will use smaller portions of
the data to speed up processing.

### PFM_loader
The first passage is to upload on R the Position frequency matrices (PFMs) with 
the specific function.

You can load PFMs either from a local file in MEME format or by specifying a 
taxonomy ID to retrieve motifs from an online database.

In the following example we are loading the first ten motifs from the local file `SwissRegulon_e_coli.meme` in the `extdata` directory.

```{r}
path_PFM <- system.file("extdata", "SwissRegulon_e_coli.meme", 
                        package = "MotifRegressR")
PFMs <- PFM_loader(path = path_PFM, tax_id = NULL)
PFMs_10 <- PFMs[1:10]

print(PFMs_10[7]) #to see only one PFM
```

### retrieve_logos
Then you can use the function `retrieve_logos` to retrieve the logo of a list of
motifs.

```{r}
logos <- retrieve_logos(PFMs = PFMs_10)

logos[7] #to see only one logo
```

### retrieve_frequencies
With this function you can load a fasta file and retrieve the nucleotide 
frequencies.

```{r}
path_fasta <- system.file("extdata", "GCF_000005845.2_ASM584v2_genomic.fna", 
                        package = "MotifRegressR")
background <- retrieve_frequencies(fasta = path_fasta)
print(background)
```

### PFM2PWM
Now you need to convert all your PFMs into Position weight matrices (PWMs).

```{r}
PWMs <- PFM2PWM(PFMs = PFMs_10, background = background)

print(PWMs[7]) #to see only one PWM
```

### get_sequences
The next passage of the pipeline is to use this function to extracts DNA 
sequences based on gene regions defined in a GFF file and retrieves the 
corresponding sequences from a FASTA file.
Given that we are working with E. coli, we set `is_circular` to TRUE. 
The `upstream` value is set to 300 by default.

```{r}
path_gff <- system.file("extdata", "GCF_000005845.2_ASM584v2_genomic.gff", 
                        package = "MotifRegressR")
dataSeq <- get_sequences(gff = path_gff, fasta = path_fasta,
                         is_circular = TRUE)
dataSeq_500 <- dataSeq[1:500,]
print(dataSeq_500[1,])
```

### scorer
Now that you have the PWMs and the gene sequences, you can calculate the scores 
by evaluating how well each sequence aligns with the position weight matrices. 
This process involves scoring each sequence based on its correspondence with the
matrix, taking into account both strands of the DNA if specified by the 
parameter `both`, and selecting the maximum score for each sequence across
the two strands.
You can modify the number of cores to use with the parameter `workers`.

```{r}
S <- scorer(PWMs = PWMs, dataSeq = dataSeq_500, workers = 4, both = TRUE)

#print(S[,1]) # to see the scores computed for the first motif
```

### compendium_loader
At this point you can load your gene expressions data.
This function loads data from a specified file path and returns it as a data
frame. It uses `read.table()` to read the file, assuming a comma-separated
values (CSV) format.

```{r}
path_compendium <- system.file("extdata", "compendium.csv", 
                               package = "MotifRegressR")
compendium <- compendium_loader(path = path_compendium, sep = ",", header = TRUE,
                                row.names = 1)

compendium_500 <- compendium[1:500, 1:10]

```


### retrieve_metadata
Now you can retrieve metadata about your compendium using this function that 
uses the sample IDs to retrieve them from the ENA browser.

```{r}
metadata <- retrieve_metadata(compendium = compendium_500)

print(metadata[1,]) #metadata of the first condition of the compendium
```

### plot_dimensionality_reduction
Here you can use this function to perform dimensionality reduction on gene 
expression data and visualizes the results using `plotly`, integrating metadata 
information.
It supports multiple methods, here we use the PCA method which is the default.

```{r}
plot_dimensionality_reduction(compendium = compendium_500, metadata = metadata)
```

### regressor and retrieve_params_list
At this point you can perform the regression. Remember that here only a small
portion of data is used as examples.
This function trains multiple regression models on the provided scores and
compendium data. It filters the rows to retain only the common ones between
the scores and the compendium before training the models. The function allows
for training various types of regression models based on the specified 
regression types and their respective parameters.

Before running the `regressor` function you need to run the `retrieve_params_list` 
function to get the `regression_params_list` parameter.

```{r}
regression_params_list <- retrieve_params_list()
regressions <- c("LM", "RF", "SVM")
models <- regressor(scores = S, compendium = compendium_500, 
                    conditions = 'all', regression_type = regressions, 
                    regression_params_list = regression_params_list)
```

### models2dataframe
With this function you can convert the previous models into dataframes.

```{r}
dfs <- models2dataframe(regression_list = models, alpha = 0.05)

```

### plot_digraph
Whit this function you can visualizes relationships between motifs and 
conditions in a bipartite directed graph. The edges represent the importance of 
connections (e.g., regression coefficients). The graph is rendered 
interactively using Plotly. Here we chose Random Forest as example.

```{r fig.width=7, fig.height=7}

plot_digraph(motif_names = colnames(S), edges_sublist = dfs["RF"])

```

### plot_clustermap
You can also build a clustermap using a sublist coming from `models2dataframe`.
Here we chose Support vector machine as example.

```{r fig.width=11, fig.height=10}
plot_clustermap(df_sublist = dfs["SVM"], distance = c("euclidean", "euclidean"), method = c("single", "single"))

```

### consensus_motifs
This function calculates the consensus motifs across a list of data frames.

```{r}
consensus <- consensus_motifs(df_list = dfs, by_regressor = TRUE)

```

### consensus_motifs_heatmap
Then you can create a heatmap to visualize the inclusivity of motifs across 
conditions or regression models based on the previous output.

```{r}
consensus_motifs_heatmap(consensus_output = consensus)
```

## Optional steps
Here there are some additional passages that you can do before the regression

### principal_comps
If you have a large dataframe, for example from your compendium data(?), you can apply principal component analysis to the matrix and use the reduced matrix as input for the regression function. For this purpose you can use the `principal_comps` function. Here, we left all the default parameters unchanged.

```{r example-chunk, eval=FALSE}
principal_comps(matrix = S, explained_var = 95, axis = 2, scale = TRUE)
```

### table_first_genes
?

```{r}
TSV <- system.file("extdata", "TUSet.tsv", package = "MotifRegressR")
PRO <- system.file("extdata", "511145_protein_info_v12_0.txt", package = "MotifRegressR")

first_genes <- table_first_genes(tsv = TSV, pro = PRO) 
```

### filter
?

```{r}
S_filtered <- filter(to_filter = S, to_keep = first_genes)
compendium_filtered <- filter(to_filter = compendium_500, to_keep = first_genes)

```

## Web interface app
mettere link per l'app

mettere screen per spiegare come si usa

Here you can find a video with all the explanations about the application:

<div align="center">
  <video width="640" height="360" controls>
    <source src="`r system.file('exdata', 'videoSpiegazione.mp4', package = 'MotifRegressR')`" type="video/mp4">
    Il tuo browser non supporta il video.
  </video>
</div>

